### PR TITLE
Speedy Compiler: factorize and clean code

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
We factorize scala code in the compiler:
 
Here are some example of changes:
+ factorize the common part of `compileChoice` and `compileChoiceByKey`
+ factorize the common part of `compileFetch` and `compileFetchByKey`
+ application patterns like 
   ```scala
    SEApp(expr1, Array(expr2))
   ```
  are replaced by: 
   ```scala
   app(expr1, expr2)
    ``` 
+ let binding patterns like  
  ```scala 
      val x1 = expr1
      val x1Pos = nextPosition()
      SLet(x1, ...)
   ```
    are replaced by pattern 
  ```scala 
   let(expr){ x1Pos => ... }
  ```
+ function binding pattern like 
  ```scala 
      Abs(1){
        val x1Pos = nextPosition()
        x1
      }
   ```
    are replaced by 
   ```scala
   function{ x1Pos => ... }
   ```

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by hurryabit — APPROVED** [review]
That looks reasonable. Thank you.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 1]
```diff
@@ -211,6 +206,37 @@ private[lf] final class Compiler(
       case None => expr
     }
 
+  @inline
+  private[this] def app(f: SExpr, a: SExpr) = SEApp(f, Array(a))
+
+  @inline
+  private[this] def let(bound: SExpr)(body: Position => SExpr) =
+    SELet1General(bound, body(nextPosition()))
```
Comment (nickchapman-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:214]: Is it better to call `SELet1` ?

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 2]
```diff
@@ -146,39 +143,37 @@ private[lf] final class Compiler(
...
     env.varIndices.get(varRef).map(svar)
 
-  def lookupExprVar(name: ExprVarName): SEVar =
+  @inline
+  private[this] def lookupExprVar(name: ExprVarName): SEVar =
     lookupVar(EVarRef(name))
       .getOrElse(throw CompilationError(s"Unknown variable: $name. Known: ${vars.mkString(",")}"))
 
-  def lookupTypeVar(name: TypeVarName): Option[SEVar] =
+  @inline
```
Comment (nickchapman-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:175]: I'm always dubious about the value of adding `@inline` especially as we have no perf stats to indicate they actually have any benefit, or even that they don't harm. We just seem to add them sometimes when we think the methods look _small enough_. But surely the JVM compiler would be the better judge of that :shrug:

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 3]
```diff
@@ -790,53 +789,47 @@ private[lf] final class Compiler(
...
         }
-      )
+      }
     }
 
   @inline
   private[this] def compileGetParty(expr: Expr): SExpr =
-    function(1, "getPArty") {
-      case List(tokenPos) => SBSGetParty(compile(expr), svar(tokenPos))
+    labeledFunction("getPArty") { tokenPos =>
```
Comment (nickchapman-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:804]: `"getPArty"` with a capital `A` !! (I realise it was like this before)

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 4]
```diff
@@ -759,28 +769,17 @@ private[lf] final class Compiler(
...
-              beginCommit,
-              result
-            ) in
-              SBSEndCommit(mustFail = false)(svar(resultPos), svar(tokenPos))
+      let(compile(partyE)) { partyPos =>
+        let(compile(updateE)) { updatePos =>
+          labeledFunction("submit") { tokenPos =>
+            let(SBSBeginCommit(optLoc)(svar(partyPos), svar(tokenPos))) { _ =>
+              let(app(svar(updatePos), svar(tokenPos))) { resultPos =>
+                SBSEndCommit(mustFail = false)(svar(resultPos), svar(tokenPos))
```
Comment (nickchapman-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:777]: I really like the look of this! :+1:

**Review by nickchapman-da — APPROVED** [review]

**Comment by nickchapman-da** [comment]
This looks like a really nice refactoring. Thanks!

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala (hunk 1 above)
Comment (dajmaki) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:214]: Do we now use `SELet` anywhere? Can we drop it?

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 5]
```diff
@@ -211,6 +206,37 @@ private[lf] final class Compiler(
...
 
+  @inline
+  private[this] def app(f: SExpr, a: SExpr) = SEApp(f, Array(a))
+
+  @inline
+  private[this] def let(bound: SExpr)(body: Position => SExpr) =
+    SELet1General(bound, body(nextPosition()))
+
+  @inline
+  private[this] def function(body: Position => SExpr): SExpr =
```
Comment (dajmaki) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:217]: would `unaryFunction` be a clearer name?

**Review by dajmaki — APPROVED** [review]
Looks good, but scary! :)

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:203]: Why not `SELet` ? in any case on another PR.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala]: I remove the one I added.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala (hunk 3 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala]: good catch!

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:203]: @dajmaki we will drop the SELet in another PR.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala (hunk 5 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:205]: changed.

Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:205]: > Looks good, but scary! :)

To be really scared you have to consider this changes together with  #7492, #7502, #7472 
;)

**Pull request merged**